edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,131 @@
+use futures::executor::block_on;
+use futures::join;
+use nix::sys::signal::{self, Signal};
+use nix::unistd::Pid;
+use signal_hook::{iterator::Signals, SIGINT, SIGTERM};
+use std::io::Error;
+use std::process::{exit, Child, Command, Stdio};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<(), Error> {
+    block_on(async_main())
+}
+
+async fn async_main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_sigterm_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!("Failed to register sigterm listener. Shutting down.");
+            return Err(e);
+        }
+    };
+
+    // start edgehub and blow up if can't start
+    let mut edgehub = Command::new("dotnet")
+        .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+        .stdout(Stdio::inherit())
+        .spawn()
+        .expect("Failed to start Edge Hub process");
+    info!("Launched Edge Hub process with pid {:?}", edgehub.id());
+
+    // unwrap broker if started, else shutdown edgehub and exit
+    let maybe_broker = match Command::new("/usr/local/bin/mqttd")
+        .stdout(Stdio::inherit())
+        .spawn()
+    {
+        Ok(child) => Some(child),
+        Err(e) => {
+            error!("Failed to start MQTT Broker process. {:?}", e);
+            None
+        }
+    };
+    let mut broker = maybe_broker.unwrap_or_else(|| {
+        info!("Broker process not started, so shutting down EdgeHub and exiting");
+        let shutdown = shutdown(&mut edgehub);
+        block_on(shutdown);
+        exit(1);
+    });
+    info!("Launched MQTT Broker process");
+
+    while !should_shutdown.load(Ordering::Relaxed) {
+        if !is_running(&mut edgehub) {
+            break;
+        }
+
+        if !is_running(&mut broker) {
+            break;
+        }
+
+        thread::sleep(Duration::from_secs(1));
+    }
+
+    let shutdown_edgehub = shutdown(&mut edgehub);
+    let shutdown_broker = shutdown(&mut broker);
+    join!(shutdown_edgehub, shutdown_broker);
```

Comment:
async by itself doesn't make function to run in a separate task. right now code execute separately but will block threads and not tasks. more changes required if you want to implement it properly.

Reply:
You can explain this when we pair program.